fl4health/model_bases/masked_layers.py

```diff
@@ -0,0 +1,359 @@
+import copy
+from typing import Optional, Union
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from torch import Tensor
+from torch.nn.common_types import _size_1_t, _size_2_t, _size_3_t
+from torch.nn.modules.utils import _pair, _single, _triple
+from torch.nn.parameter import Parameter
+
+from fl4health.utils.functions import bernoulli_sample
+
+
+class MaskedLinear(nn.Linear):
+    """
+    Implementation of masked linear layers.
+
+    Like regular linear layers (i.e., nn.Linear module), a masked linear layer has a weight and a bias.
+    However, the weight and the bias do not receive gradient in back propagation.
+    Instead, two score tensors - one for the weight and another for the bias - are maintained.
+    In the forward pass, the score tensors are transformed by the Sigmoid function into probability scores,
+    which are then used to produce binary masks via bernoulli sampling.
+    Finally, the binary masks are  applied to the weight and the bias. During training,
+    gradients with respect to the score tensors are computed and used to update the score tensors.
+
+    Args:
+        in_features: size of each input sample
+        out_features: size of each output sample
+        bias: If set to ``False``, the layer will not learn an additive bias.
+            Default: ``True``
+
+    Attributes:
+        weight: weights of the module.
+        bias:  bias of the module.
+        weight_score: learnable scores for the weights. Has the same shape as weight.
+        bias_score: learnable scores for the bias. Has the same shape as bias.
+    """
+
+    __constants__ = ["in_features", "out_features"]
+    in_features: int
```

Comment:
They were declared in `nn.Linear` in this way so I just followed it without thinking too much. But after looking into it this is what I've found:

1. `__constants__` is an annotation from [TorchScript](https://pytorch.org/docs/stable/jit.html#constants) that marks certain members of a class as fixed. This can inform the PyTorch JIT (Just-In-Time) compiler to make certain assumptions for optimization purposes.
2. I couldn't find out the reason for declaring `weight: Tensor` before `__init__`. Normally the fields declared before `__init__` are static attributes belonging to the class rather than instance attributes, but here we declare `weight` as an instance attribute afterwards anyway, so maybe it was done for type annotation and better readability?

In any case I'm removing these declarations because they have already been done in the parent class

Reply:
Sounds good